File: elixir/lib/sensedata_web/schema.ex
defmodule SensedataWeb.Schema do
  use Absinthe.Schema

  object :image do
    field :href, :string
    field :title, :string
  end

  import_types SensedataWeb.Schema.Viewer
  import_types SensedataWeb.Schema.Contract

  query do
    import_fields :viewer_queries
    import_fields :contract_queries
  end
end

File: go/src/sensedata/schema.go
package sensedata

const Schema = `
schema {
  query: Query
}

scalar Date

type Query {
  viewer: Viewer
  contracts(filter: ContractFilter): PaginatedContracts
}

type Image {
  href: String!
  title: String!
}

type Viewer {
  name: String!
  avatar: Image!
  notifications: [Notification!]!
}

enum NotificationKind { WARNING, ERROR, INFO }

type Notification {
  title: String!
  kind: NotificationKind!
}

enum ListFilterOuter { ALL, PROFILE }

input ContractFilter {
  outer: ListFilterOuter
}

type PaginatedContracts {
  totalCount: Int!
  nodes: [Contract!]!
}

type Contract {
  amount: Float!
  originalId: String!
}
`
